homeassistant/components/backup/config.py

```diff
@@ -271,16 +338,81 @@ def _delete_filter(
+                    for agent_retention in agents_retention.values()
+                )
+                if (global_days := self.days) is None and not has_agents_retention_days:
+                    # No global retention days and no agent retention days
                     return {}
+
                 now = dt_util.utcnow()
-                return {
-                    backup_id: backup
-                    for backup_id, backup in backups.items()
-                    if dt_util.parse_datetime(backup.date, raise_on_error=True)
-                    + timedelta(days=self.days)
-                    < now
-                }
+                if global_days is not None and not has_agents_retention:
+                    # Return early to avoid the longer filtering below.
+                    return {
+                        backup_id: backup
+                        for backup_id, backup in backups.items()
+                        if dt_util.parse_datetime(backup.date, raise_on_error=True)
+                        + timedelta(days=global_days)
+                        < now
+                    }
+
+                # If there are any agent retention settings, we need to check
+                # the retention settings, for every backup and agent combination.
+
+                backups_to_delete = {}
+
+                for backup_id, backup in backups.items():
+                    backup_date = dt_util.parse_datetime(
+                        backup.date, raise_on_error=True
+                    )
+                    agents_for_backup_to_delete = set(backup.agents)
+                    for agent_id in backup.agents:
+                        agent_retention = agents_retention.get(agent_id)
+                        if agent_retention is None:
+                            # This agent does not have a retention setting,
+                            # so the global retention setting should be used.
+                            if global_days is None:
+                                # This agent does not have a retention setting
+                                # and the global retention days setting is None,
+                                # so this backup should not be deleted.
+                                agents_for_backup_to_delete.discard(agent_id)
+                                continue
+                            days = global_days
+                        elif (agent_days := agent_retention.days) is None:
+                            # This agent has a retention setting
+                            # where days is set to None,
+                            # so the backup should not be deleted.
+                            agents_for_backup_to_delete.discard(agent_id)
+                            continue
+                        else:
+                            # This agent has a retention setting
+                            # where days is set to a number,
+                            # so the backup should be deleted.
```

Comment:
The comment is misleading, we don't delete here, we just update `days` according to the agent setting